src/v/cluster/partition_manager.cc

```diff
@@ -297,6 +297,12 @@ ss::future<consensus_ptr> partition_manager::manage(
 
     co_await p->start(_stm_registry, xst_state);
 
+    // this is not done in partition::start itself because the purpose of this
+    // flag is to operate in an uninterruptible context with watcher
+    // notification below to avoid registration while this fiber is blocked,
+    // leading to double notifications.
+    p->mark_started();
```

Comment:
I think this is leaking the abstraction into the manager.  Generally speaking (outside the context this bug), having a field like "_started" and setting it within the partition::start() itself seems like a common pattern and in this case it so happens that the notification mechanism depends on it, setting it from the outside seems a little awkward at first glance.

Reply:
> having a field like "_started" and setting it within the partition::start()

In the following

```
ss::future<> partition_manager::manage() {
  ...
  co_await p->start();
  watcher.notify(p);
}

ss::future<> partition::start() {
  co_await do_start();
  _started = true;
}

ss::future<> partition::do_start() {
  // read method
}
```

how do i know that `_started = true` is executing in an uninterruptible context with the caller (which is required to prevent the race we care about)?

it's true that seastar does lazy suspend (no scheduling point entering a coroutine), but i don't know if there is a similar guarantee on return.

that's the motivation for `mark_started`

Reply:
> which is required to prevent the race we care about?

yes you are right. I missed the "uninterruptible context" part in the comment.